tensorflow/lite/micro/micro_profiler.h

```diff
@@ -33,67 +47,174 @@ class MicroProfiler : public MicroProfilerInterface {
+    TFLITE_DCHECK(event_handle < MAX_EVENTS);
+    end_ticks_[event_handle] = GetCurrentTimeTicks();
+  }
 
   // Clears all the events that have been currently profiled.
-  void ClearEvents() { num_events_ = 0; }
+  void ClearEvents() {
+    num_events_ = 0;
+    num_tag_groups_ = 0;
+  }
 
   // Returns the sum of the ticks taken across all the events. This number
   // is only meaningful if all of the events are disjoint (the end time of
   // event[i] <= start time of event[i+1]).
-  uint32_t GetTotalTicks() const;
+  uint32_t GetTotalTicks() const {
+    int32_t ticks = 0;
+    for (int i = 0; i < num_events_; ++i) {
+      ticks += end_ticks_[i] - start_ticks_[i];
+    }
+    return ticks;
+  }
+
+  // Prints the profiling information of each of the events with the
+  // given format (human readable by default).
+  void Log(MicroProfilerLogFormat format =
+               MicroProfilerLogFormat::HumanReadable) const {
+#if !defined(TF_LITE_STRIP_ERROR_STRINGS)
+    switch (format) {
+      case MicroProfilerLogFormat::HumanReadable:
+        for (int i = 0; i < num_events_; ++i) {
+          uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+          uint64_t us = TicksToUs(ticks);
+          MicroPrintf("%s took %" PRIu64 ".%" PRIu64 " ms (%u ticks)", tags_[i],
+                      us / 1000, us % 1000, ticks);
+        }
+        break;
+
+      case MicroProfilerLogFormat::Csv:
+        MicroPrintf("\"Event\",\"Tag\",\"Ticks\"");
+        for (int i = 0; i < num_events_; ++i) {
+#if defined(HEXAGON) || defined(CMSIS_NN)
+          int ticks = end_ticks_[i] - start_ticks_[i];
+          MicroPrintf("%d,%s,%d", i, tags_[i], ticks);
+#else
+          uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+          MicroPrintf("%d,%s,%" PRIu32, i, tags_[i], ticks);
+#endif
+        }
+        break;
+    }
+#endif  // !defined(TF_LITE_STRIP_ERROR_STRINGS)
+  }
+
+  // Prints the profiling information of each of the events in CSV format.
+  void LogCsv() const { Log(MicroProfilerLogFormat::Csv); }
 
   // Prints the profiling information of each of the events in human readable
-  // form.
-  void Log() const;
+  // form, grouped per tag, sorted by execution time.
+  void LogGrouped(MicroProfilerLogFormat format) {
+#if !defined(TF_LITE_STRIP_ERROR_STRINGS)
+    for (int i = 0; i < num_events_; ++i) {
+      // Find if the tag already exists in uniqueTags
+      TagGroup& tag_group = GetTagGroup(tags_[i]);
 
-  // Prints the profiling information of each of the events in CSV (Comma
-  // Separated Value) form.
-  void LogCsv() const;
+      uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+      tag_group.tag = tags_[i];
+      tag_group.ticks += ticks;
+      tag_group.tag_count++;
+    }
 
-  // Prints  total ticks for each unique tag in CSV format.
-  // Output will have one row for each unique tag along with the
-  // total ticks summed across all events with that particular tag.
-  void LogTicksPerTagCsv();
```

Comment:
This method cannot be removed as it breaks existing applications.

Reply:
Added it back

Reply:
Is there a policy to _never_ break stuff like this? IMO duplicating APIs like this is also not great.

Reply:
@andresovela Let's just say, that at the present time, no one is allowed to break anything. :slightly_smiling_face: